Updated the code that allows to save the cropped image in result_img directory

shubham-shahh opened the pull request:
The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' and the code to save the image in result_img dir isn't updated based on the above conversion which was causing errors while recompiling the code. 

So, I've updated that, and now one can store the detected images in result_img dir. I've also updated the Readme.md based on that

cenit commented on README.md:
```diff
@@ -288,6 +288,7 @@ On Linux find executable file `./darknet` in the root directory, while on Window
 * To calculate anchors: `darknet.exe detector calc_anchors data/obj.data -num_of_clusters 9 -width 416 -height 416`
 * To check accuracy mAP@IoU=50: `darknet.exe detector map data/obj.data yolo-obj.cfg backup\yolo-obj_7000.weights`
 * To check accuracy mAP@IoU=75: `darknet.exe detector map data/obj.data yolo-obj.cfg backup\yolo-obj_7000.weights -iou_thresh 0.75`
+* To save the cropped image of the object detected: [`uncomment the following lines of code`](https://github.com/shubham-shahh/darknet/blob/95ec9bd9249a7c79bf2fea06dee964365f2fb61b/src/image_opencv.cpp#L980-L996)
```
this line in readme is pointing to your repo code. This has to point to alexeyab repo, even if the link is not going to work until merged with master

shubham-shahh replied:
I'll update that right away. thanks

shubham-shahh replied:
> this line in readme is pointing to your repo code. This has to point to alexeyab repo, even if the link is not going to work until merged with master

Updated the link as mentioned.

cenit replied:
yes but pointing to a specific commit sha might not be the best option. Better pointing to the master trunk.

shubham-shahh replied:
Yes, makes sense. I committed the changes you mentioned. thanks once again

cenit commented:
I have updated your initial message and reopened initial issue. The magic token "Fixes: #issue" will close the issue if/when this PR is merged automatically

shubham-shahh commented:
> I have updated your initial message and reopened initial issue. The magic token "Fixes: #issue" will close the issue if/when this PR is merged automatically

okay, thanks :)

cenit commented on README.md:
```diff
@@ -288,7 +288,7 @@ On Linux find executable file `./darknet` in the root directory, while on Window
 * To check accuracy mAP@IoU=50: `darknet.exe detector map data/obj.data yolo-obj.cfg backup\yolo-obj_7000.weights`
 * To check accuracy mAP@IoU=75: `darknet.exe detector map data/obj.data yolo-obj.cfg backup\yolo-obj_7000.weights -iou_thresh 0.75`
-* To save the cropped image of the object detected: [`uncomment the following lines of code`](https://github.com/shubham-shahh/darknet/blob/95ec9bd9249a7c79bf2fea06dee964365f2fb61b/src/image_opencv.cpp#L980-L996)
+* To save the cropped image of the object detected: [`uncomment the following lines of code`](https://github.com/AlexeyAB/darknet/blob/64efa721ede91cd8ccc18257f98eeba43b73a6af/src/image_opencv.cpp#L980-L994)
```
```suggestion
* To save the cropped image of the object detected: [`uncomment the following lines of code`](https://github.com/AlexeyAB/darknet/blob/master/src/image_opencv.cpp#L980-L996)
```

shubham-shahh commented:
> I have updated your initial message and reopened initial issue. The magic token "Fixes: #issue" will close the issue if/when this PR is merged automatically

Are we merging the PR with the Master branch?

cenit commented:
@AlexeyAB what's your opinion on this PR? It fixes lines of code that are commented, as such they do not break nor fix anything. But, if the user was to uncomment those lines, they were broken and now ok

shubham-shahh commented:
> @AlexeyAB what's your opinion on this PR? It fixes lines of code that are commented, as such they do not break nor fix anything. But, if the user was to uncomment those lines, they were broken and now ok

Any Update @AlexeyAB @cenit

shubham-shahh commented:
> 1. The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' and the code to save the image in result_img dir isn't
>    updated based on the above conversion which was causing errors while recompiling the code.
>    So, I've updated that, and now one can store the detected images in result_img dir. I've also updated the Readme.md based
>    on that
>    Fixes: #7276
> 2. I've added a new feature with which a user can run inference on all the images in the mentioned folder. I've also updated the readme with the information on how to use this feature.

@cenit sir please review my recent changes, in which I've added this feature.
Thanks.

shubham-shahh commented on README.md:
```diff
@@ -288,6 +288,8 @@ On Linux find executable file `./darknet` in the root directory, while on Window
 * To check accuracy mAP@IoU=50: `darknet.exe detector map data/obj.data yolo-obj.cfg backup\yolo-obj_7000.weights`
 * To check accuracy mAP@IoU=75: `darknet.exe detector map data/obj.data yolo-obj.cfg backup\yolo-obj_7000.weights -iou_thresh 0.75`
+* To save the cropped image of the object detected: [`uncomment the following lines of code`](https://github.com/AlexeyAB/darknet/blob/master/src/image_opencv.cpp#L980-L996)
+* To run inference on all the images in a folder: `./darknet detector test ./cfg/coco.data ./cfg/yolov4.cfg ./yolov4.weights -i 0 -thresh 0.25 --folder_inference ./result_img/` here `./result_img/` is the path of the folder in which the inference has to be done. the folder name can be anything, as long as it exists. After the inference the images will be moved to `./processed_files/`. To change t ...
```
updated the readme.md

cenit replied:
Ad always, please use links to alexeyab's repo, without any commit id reference but pointing to the latest master trunk

shubham-shahh replied:
okay, sir, I'll do that

shubham-shahh replied:
Sir, I've updated the link but it is broken until the PR is merged @cenit

cenit replied:
Yes of course link is broken until merged, but this is a requirement for being considered for merging!

shubham-shahh commented on include/darknet.h:
```diff
@@ -1040,7 +1040,7 @@ LIB_API float *network_predict_image_letterbox(network *net, image im);
 LIB_API void train_detector(char *datacfg, char *cfgfile, char *weightfile, int *gpus, int ngpus, int clear, int dont_show, int calc_map, int mjpeg_port, int show_imgs, int benchmark_layers, char* chart_path);
 LIB_API void test_detector(char *datacfg, char *cfgfile, char *weightfile, char *filename, float thresh,
-    float hier_thresh, int dont_show, int ext_output, int save_labels, char *outfile, int letter_box, int benchmark_layers);
+    float hier_thresh, int dont_show, int ext_output, int save_labels, char *outfile, int letter_box, int benchmark_layers, char *folder_inference);
```
added a new argument to the function

shubham-shahh commented on src/detector.c:
```diff
@@ -1638,110 +1648,255 @@ void test_detector(char *datacfg, char *cfgfile, char *weightfile, char *filenam
-            if (lk.type == YOLO || lk.type == GAUSSIAN_YOLO || lk.type == REGION) {
-                l = lk;
-                printf(" Detection layer: %d - type = %d \n", k, l.type);
+    if(folder_inference==NULL){
```
made changes in such a way that the original functionality is not changed by any means

shubham-shahh commented on README.md:
```diff
@@ -288,6 +288,8 @@ On Linux find executable file `./darknet` in the root directory, while on Window
 * To check accuracy mAP@IoU=50: `darknet.exe detector map data/obj.data yolo-obj.cfg backup\yolo-obj_7000.weights`
 * To check accuracy mAP@IoU=75: `darknet.exe detector map data/obj.data yolo-obj.cfg backup\yolo-obj_7000.weights -iou_thresh 0.75`
+* To save the cropped image of the object detected: [`uncomment the following lines of code`](https://github.com/AlexeyAB/darknet/blob/master/src/image_opencv.cpp#L980-L996)
+* To run inference on all the images in a folder: `./darknet detector test ./cfg/coco.data ./cfg/yolov4.cfg ./yolov4.weights -i 0 -thresh 0.25 --folder_inference ./result_img/` here `./result_img/` is the path of the folder in which the inference has to be done. the folder name can be anything, as long as it exists. After the inference the images will be moved to `./processed_files/`. To change t ...
```
Sir, I've updated the link but it is broken until the PR is merged @cenit

AlexeyAB commented:
But why is 
`./darknet detector test ./cfg/coco.data ./cfg/yolov4.cfg ./yolov4.weights -i 0 -thresh 0.25 --folder_inference ./result_img/`
 better than 
`darknet.exe detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < data/train.txt > result.txt` ?

This  `< data/train.txt` is more flexible, since you can put either all images from folder to the `data/train.txt` or only some images.

shubham-shahh commented:
> But why is
> `./darknet detector test ./cfg/coco.data ./cfg/yolov4.cfg ./yolov4.weights -i 0 -thresh 0.25 --folder_inference ./result_img/`
> better than
> `darknet.exe detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < data/train.txt > result.txt` ?
> 
> This `< data/train.txt` is more flexible, since you can put either all images from folder to the `data/train.txt` or only some images.

Hello sir, I hope you are doing well. 
`darknet.exe detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < data/train.txt > result.txt` this command works well when the user knows the list of images in a folder and if no new images are added or subtracted from the folder because the program ends when the inference on all the images is done

`./darknet detector test ./cfg/coco.data ./cfg/yolov4.cfg ./yolov4.weights -i 0 -thresh 0.25 --folder_inference ./result_img/` this command works well when new images are added constantly to a folder, as this program is non terminating and keeps looking for images in the folder. A common use case could be if new images are added to a folder from some other process and the user wants to do further inference on all the images in that folder. in this case new images are constantly added to the folder hence a user cannot make a list of it.

thanks for this awesome project @AlexeyAB

AlexeyAB commented:
I think it is not worth to complicate the code due to the functionality that is easily achieved by these functions.

* Linux:

`ls /home/user/*.jpg > filelist.txt`

`./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt`

----

* Windows:

`dir C:\images /b /s > "filelist.txt"`

`darknet.exe detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt`

shubham-shahh commented:
> I think it is not worth to complicate the code due to the functionality that is easily achieved by these functions.
> 
> * Linux:
> 
> `ls /home/user/*.jpg > filelist.txt`
> 
> `./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt`
> 
> * Windows:
> 
> `dir C:\images /b /s > "filelist.txt"`
> 
> `darknet.exe detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt`

Sir, The procedure you mentioned is still terminating and the only way I can think of to make it non terminating is by creating a shell script that creates .txt files after every regular interval or to edit the code. In this approach, the user also has to make another .txt file to keep a track of images on which the inference has been done or delete the images or move them which again requires editing the code or putting that in a shell script. with this approach, 
`./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt` has to be executed again and again which means the user has to load the model in GPU memory at regular intervals which is time-consuming.
Regarding the complexity of code, I've added a simple `if statement` if the user wants to use this functionality [this](https://github.com/shubham-shahh/darknet/blob/cb6676457f2e1d8f0cd61e0176795ad154106445/src/detector.c#L1780-L1925) block of code will execute. and if the user doesn't use the functionality the code would execute the way it used to be.

shubham-shahh commented:
Are we merging this PR? @AlexeyAB @cenit

shubham-shahh commented:
> > I think it is not worth to complicate the code due to the functionality that is easily achieved by these functions.
> > 
> > * Linux:
> > 
> > `ls /home/user/*.jpg > filelist.txt`
> > `./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt`
> > 
> > * Windows:
> > 
> > `dir C:\images /b /s > "filelist.txt"`
> > `darknet.exe detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt`
> 
> Sir, The procedure you mentioned is still terminating and the only way I can think of to make it non terminating is by creating a shell script that creates .txt files after every regular interval or to edit the code. In this approach, the user also has to make another .txt file to keep a track of images on which the inference has been done or delete the images or move them which again requires editing the code or putting that in a shell script. with this approach,
> `./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt` has to be executed again and again which means the user has to load the model in GPU memory at regular intervals which is time-consuming and resource intensive.
> Regarding the complexity of code, I've added a simple `if statement` if the user wants to use this functionality [this](https://github.com/shubham-shahh/darknet/blob/cb6676457f2e1d8f0cd61e0176795ad154106445/src/detector.c#L1780-L1925) block of code will execute. and if the user doesn't use the functionality the code would execute the way it used to be.

Hi,
Sir What's your opinion on this? @cenit @AlexeyAB

shubham-shahh commented:
Hello sir,
I am planning to add a tracker to darknet can you please point me towards some existing work on which I can build upon or I'll have to make from scratch @cenit @AlexeyAB

shubham-shahh commented:
> ./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt
> 
> NB!

What are you trying to say? Can you please clear lly describe your issue?

rickkcir commented:
> > ./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt
> > NB!
> 
> What are you trying to say? Can you please clearly describe your issue?

I want to get the coordinates,then I use this commad, but I found the file only saved relative coordinates,while I want to save the  absolute coordinates, do you know how to realize it? thanks

rg321 commented:
Hi, can you try following code
in file src/image_opencv.cpp, show_img is Mat pointer. 
Just crop desired portion out of Mat.

`// you should create directory: result_img
 static int copied_frame_id = -1;
 // static IplImage* copy_img = NULL;
 if (copied_frame_id != frame_id) {
    copied_frame_id = frame_id;
    // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels);
    // cvCopy(show_img, copy_img, 0);
 }
 static int img_id = 0;
 img_id++;
 char image_name[1024];
 sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]);
 CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y);
 cv::Mat copy_image = (*show_img)(rect);
 cv::imwrite(image_name, copy_image);
 // cvSetImageROI(copy_img, rect);
 // cvSaveImage(image_name, copy_img, 0);
 // cvResetImageROI(copy_img);`

pull request created for the same
[https://github.com/AlexeyAB/darknet/pull/8197](https://github.com/AlexeyAB/darknet/pull/8197)

shubham-shahh commented:
> > > ./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights -ext_output -dont_show -out result.json < filelist.txt > result.txt
> > > NB!
> > 
> > 
> > What are you trying to say? Can you please clearly describe your issue?
> 
> I want to get the coordinates,then I use this commad, but I found the file only saved relative coordinates,while I want to save the absolute coordinates, do you know how to realize it? thanks

You can get the absolute co ordinates since you already know the original image size

shubham-shahh commented:
> Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> 
> `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> 
> pull request created for the same #8197

Can you clearly mention what exactly you're expecting?

Thanks

shubham-shahh reviewed (commented):
@cenit what does this commit do?

rg321 commented:
> > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > pull request created for the same #8197
> 
> Can you clearly mention what exactly you're expecting?
> 
> Thanks

@shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of  image. I just provided a simple way of doing that. As you pointed out, "The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat'", so just crop the detected area out of original image. For that I have just added following single line -:
`cv::Mat copy_image = (*show_img)(rect)`
rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -:
`cv::imwrite(image_name, copy_image)`
Thats it, problem solved. You don't have to worry that "'show_img' has been changed from 'iplimage' to 'cv::Mat'".

shubham-shahh commented on src/image_opencv.cpp:
```diff
@@ -997,12 +997,14 @@ extern "C" void draw_detections_cv_v3(mat_cv* mat, detection *dets, int num, flo
-                //    if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels);
-                //    cvCopy(show_img, copy_img, 0);
+                //    if(copy_img == NULL) copy_img = cvCreateImage(cvSize(ipltypeimg->width, ipltypeimg->height), ipltypeimg->depth, ipltypeimg->nChannels);
+                //    cvCopy(ipltypeimg, copy_img, 0);
```
It is solved here @rg321

shubham-shahh commented:
> > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > pull request created for the same #8197
> > 
> > 
> > Can you clearly mention what exactly you're expecting?
> > Thanks
> 
> @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".
 

> > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > pull request created for the same #8197
> > 
> > 
> > Can you clearly mention what exactly you're expecting?
> > Thanks
> 
> @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".

H

> > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > pull request created for the same #8197
> > 
> > 
> > Can you clearly mention what exactly you're expecting?
> > Thanks
> 
> @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".

Hi, this PR also implements some more features along with solving that bug. That bug is simply solved by converting the image to ipltype, I tagged you where I have done that conversion. Rest changes of this PR are made to implement several other features

Thanks

rg321 commented:
> > > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > > pull request created for the same #8197
> > > 
> > > 
> > > Can you clearly mention what exactly you're expecting?
> > > Thanks
> > 
> > 
> > @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".
> 
> > > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > > pull request created for the same #8197
> > > 
> > > 
> > > Can you clearly mention what exactly you're expecting?
> > > Thanks
> > 
> > 
> > @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".
> 
> H
> 
> > > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > > pull request created for the same #8197
> > > 
> > > 
> > > Can you clearly mention what exactly you're expecting?
> > > Thanks
> > 
> > 
> > @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".
> 
> Hi, this PR also implements some more features along with solving that bug. That bug is simply solved by converting the image to ipltype, I tagged you where I have done that conversion. Rest changes of this PR are made to implement several other features
> 
> Thanks

In that case, you may either exactly mention in title of this pull request or you may want to create different pull request for the bug and a different one for features implemented. That would provide more clarity to your pull request. If possible, provide appropriate title to your pull request, what bug it solves or what feature it implements and add changes necessary for that only in that particular pull request. Else, as I mentioned above, for the problem mentioned in the title, my changes are good enough.

shubham-shahh commented:
> > > > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > > > pull request created for the same #8197
> > > > 
> > > > 
> > > > Can you clearly mention what exactly you're expecting?
> > > > Thanks
> > > 
> > > 
> > > @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".
> > 
> > 
> > > > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > > > pull request created for the same #8197
> > > > 
> > > > 
> > > > Can you clearly mention what exactly you're expecting?
> > > > Thanks
> > > 
> > > 
> > > @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".
> > 
> > 
> > H
> > > > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > > > pull request created for the same #8197
> > > > 
> > > > 
> > > > Can you clearly mention what exactly you're expecting?
> > > > Thanks
> > > 
> > > 
> > > @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".
> > 
> > i

> > > > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > > > pull request created for the same #8197
> > > > 
> > > > 
> > > > Can you clearly mention what exactly you're expecting?
> > > > Thanks
> > > 
> > > 
> > > @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".
> > 
> > 
> > > > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > > > pull request created for the same #8197
> > > > 
> > > > 
> > > > Can you clearly mention what exactly you're expecting?
> > > > Thanks
> > > 
> > > 
> > > @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".
> > 
> > 
> > H
> > > > > Hi, can you try following code in file src/image_opencv.cpp, show_img is Mat pointer. Just crop desired portion out of Mat.
> > > > > `// you should create directory: result_img static int copied_frame_id = -1; // static IplImage* copy_img = NULL; if (copied_frame_id != frame_id) { copied_frame_id = frame_id; // if(copy_img == NULL) copy_img = cvCreateImage(cvSize(show_img->width, show_img->height), show_img->depth, show_img->nChannels); // cvCopy(show_img, copy_img, 0); } static int img_id = 0; img_id++; char image_name[1024]; sprintf(image_name, "result_img/img_%d_%d_%d_%s.jpg", frame_id, img_id, class_id, names[class_id]); CvRect rect = cvRect(pt1.x, pt1.y, pt2.x - pt1.x, pt2.y - pt1.y); cv::Mat copy_image = (*show_img)(rect); cv::imwrite(image_name, copy_image); // cvSetImageROI(copy_img, rect); // cvSaveImage(image_name, copy_img, 0); // cvResetImageROI(copy_img);`
> > > > > pull request created for the same #8197
> > > > 
> > > > 
> > > > Can you clearly mention what exactly you're expecting?
> > > > Thanks
> > > 
> > > 
> > > @shubham-shahh as you mentioned at top of this page, you are trying to crop detections out of image. I just provided a simple way of doing that. As you pointed out, " The type of 'show_img' has been changed from 'iplimage' to 'cv::Mat' ", so just crop the detected area out of original image. For that I have just added following single line -: `cv::Mat copy_image = (*show_img)(rect)` rect is the rectange of detected object. We just crop that area from show_img and name it copy_image. Now save copy_image and it is cropped detectetion out of original image. To save copy_image, I use following code -: `cv::imwrite(image_name, copy_image)` Thats it, problem solved. You don't have to worry that " 'show_img' has been changed from 'iplimage' to 'cv::Mat' ".
> > 
> > 
> > Hi, this PR also implements some more features along with solving that bug. That bug is simply solved by converting the image to ipltype, I tagged you where I have done that conversion. Rest changes of this PR are made to implement several other features
> > Thanks
> 
> In that case, you may either exactly mention in title of this pull request or you may want to create different pull request for the bug and a different one for features implemented. That would provide more clarity to your pull request. If possible, provide appropriate title to your pull request, what bug it solves or what feature it implements and add changes necessary for that only in that particular pull request. Else, as I mentioned above, for the problem mentioned in the title, my changes are good enough.

Hi,
you can check the description, I have clearly mentioned what bug it solves and which feature it implements

thanks

cenit commented on src/detector.c:
```diff
@@ -8,6 +8,7 @@
 #include "box.h"
 #include "demo.h"
 #include "option_list.h"
+#include <dirent.h>
```
PR is using dirent.h and code that is not cross-compatible, so is unacceptable as is

shubham-shahh replied:
I can add alternatives to dirent

shubham-shahh replied:
> 

can you please tag what part of the code which is causing issue? I will work on it. plus this is a really old PR it needs rebasing as well

cenit replied:
I rebased it on latest master for you before commenting the issue

cenit replied:
just rewrite the code in this file avoiding dirent.h inclusion (not available on windows)

shubham-shahh replied:
okay, thanks for pointing out the issues, I'll work on them

cenit commented on src/detector.c:
```diff
@@ -1642,110 +1652,255 @@ void test_detector(char *datacfg, char *cfgfile, char *weightfile, char *filenam
+        free_network(net);
     }
+    else{
+        DIR *folder;
+        struct dirent *entry;
```
these two lines, for example, depend on dirent.h being included, which is incompatible with our aim of being 100% windows compatible

shubham-shahh replied:
okay, I noted it, I'll apply required changes

shubham-shahh replied:
what's your opinion for [Boost filesystem](https://www.boost.org/doc/libs/1_43_0/libs/filesystem/doc/index.htm)? I don't have much experience in building projects in windows. all your help is appreciated

cenit replied:
if you can use std::filesystem it would be much better, so that we can avoid depending on another external lib

shubham-shahh replied:
Okay, I'll look into it
thanks

shubham-shahh replied:
Hi, @cenit I was planning to fix this PR, I just wanted to know if there are any other changes to make other than filesystem?

shubham-shahh commented:
Hi @cenit looks like you've fixed the issue related to building in windows, can we merge this PR into master?
